Fix bug for integer compressed images with blanks

rlwastro opened the pull request:
### Description

This pull request is to address a bug that raises an exception when some compressed images are read.

In `_tiled_compression.py:decompress_image_data_section()`, blanked pixels are assigned the value `np.nan`.  That is incorrect for integer datatypes and raises an exception because `np.nan` cannot be converted to an integer.

The fix is to use the `BLANK` keyword from the header.  If no `BLANK` or `ZBLANK` keyword is found for integer datatypes, the most negative integer (as determined by `ZBITPIX`) is used as a fill value. That should not happen if pixel-blanking is specified (otherwise why is `ZBLANK` there at all?). An `AstropyUserWarning` is issued in that case. 

This also fixes a bug in `decompress_image_data_section()` for the case where `BLANK` is specified in the header but `ZBLANK` is not.  The FITS standard recommends that `BLANK` should be used instead of `ZBLANK` for integer images.  For images with `BLANK` but not `ZBLANK`, the current code would not work correctly. The new version handles integer compressed images that specify either `BLANK` or `ZBLANK` (or both).

In the `compressed.py` module, integer images with blanks are converted to floats with `NaN` values. Previously that conversion happened only if `BZERO/BSCALE` keywords were also specified.  The new code does the conversion even if `BZERO` and `BSCALE` are not defined. The new behavior is consistent with the handling of these keywords in uncompressed image HDUs.

Test code is included.

<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->


<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->



<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #15236

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

neutrinoceros commented:
This would supersede #15756 and fix #15236
@rlwastro please make sure formatting is green (you may comment `pre-commit.ci autofix` here or fix it locally and push), and please include a changelog entry (see `docs/changes/README.rst`)

rlwastro commented:
> This would supersede #15756 and fix #15236 @rlwastro please make sure formatting is green (you may comment `pre-commit.ci autofix` here or fix it locally and push), and please include a changelog entry (see `docs/changes/README.rst`)

I fixed the formatting and added a changelog entry.  Let me know if you see other changes that are required.

I'm working on creating a small compressed image that can be used for testing.

rlwastro commented:
@neutrinoceros I'm testing this fix on various images and am beginning to wonder whether I have the correct fix for this issue.

Currently if we read a FITS file with an _uncompressed_ integer image that has a `BLANK` keyword in the header, the image gets converted from an integer datatype to a `float32`, and the blanked pixels are set to `np.nan`.  I could argue with that behavior (perhaps a masked integer array would be better than a `float32` array).  But that is the way it acts now, and as far as I know people are OK with it.

With my version of the code, a FITS file with a _compressed_ integer image that has a `BLANK` keyword in the header does _not_ get converted to a float.  Instead it is kept as an integer array, with the blank values retained as the special integer value.

I think for consistency with the uncompressed image behavior, this should instead convert the array to `float32` and replace the blanks with `np.nan`.  If you agree, I can withdraw this pull request and submit a new one.  But let me know if you think it needs more discussion before doing that.

This was a bit tricky to discover because the Pan-STARRS images in the original bug report also have `BZERO` and `BSCALE` keywords, and it turns out they do in the end get converted to `float32` with `np.nan` for blanks.  So the change I'm proposing will in the end give the same result for the PS images.  But I'm pretty sure the new approach would be better for other images.

I do think this is important to fix soon!  Currently it simply dies with an exception when one of these images is opened.  That is clearly bad.

neutrinoceros commented:
> I think for consistency with the uncompressed image behavior, this should instead convert the array to float32 and replace the blanks with np.nan. If you agree, I can withdraw this pull request and submit a new one. But let me know if you think it needs more discussion before doing that.

I can see both ways as satisfying solutions, albeit non ideal. I think @saimn and/or @astrofrog should weigh in here.

rlwastro commented:
@neutrinoceros I figured out the right place to make this change.  The code that applies the `BSCALE/BZERO` scaling and replaces `BLANK` values with `NaN` (in `compressed.py:CompImageHDU._scale_data()`) was skipped if the `BSCALE` and `BZERO` values were omitted.  It is now executed if `BLANK` is specified (even in the absence of `BSCALE/BZERO`).

I updated the description to reflect these additional changes.  This version should have the correct behavior (as similar to uncompressed image HDUs as possible).

There is room for improvement to the handling of errors in compressed image headers.  I have not tried to make these changes because the code in `header.py` is pretty scary.  My changes do not make these problems any worse at least!
* For integer images, if the `ZBLANK` header keyword is specified but there is no `BLANK` keyword, `ZBLANK` should be copied to `BLANK` (with a warning).
* For integer images, if neither `ZBLANK` nor `BLANK` appears in the header, but there is a `ZBLANK` column in the table, then `BLANK` should be set to a default value (with a warning).

rlwastro commented:
@neutrinoceros I made some further improvements in the code that cleaned up some remaining issues with unusual file formats.  I believe this patch is complete now.

I added new tests in the `test_tiled_compression.py` module that I believe should test all the added code.  I added some additional (small) test files as well to support those tests. There are still some warnings about code coverage.  I'm not sure what the problem is there.  I can't tell whether the new tests have actually been run yet or not.

I believe this is very close to complete, so if you or others can help get it through the final stages, that would be very helpful.

rlwastro commented:
Thanks @pllim !  I fixed some filename typos in my test code, and now the patch coverage is 100% complete.  All checks now pass.

pllim commented:
Alas, we just updated branch protection rules due to matplotlib bump. Would you mind rebasing? Thanks!

rlwastro commented:
@pllim Ah crap, I think I screwed that up.  This is very much out of my comfort zone in git!  I'll see if I can fix it.  If not I may have to start over.

pllim commented:
Maybe try this?

```
git fetch upstream main
git rebase upstream/main
```

And see if the history looks better before you force push?

rlwastro commented:
Whew, thanks!  That did the trick.  I think things look like they will be OK now.  I'm just waiting to confirm the checks are all still OK.

neutrinoceros commented on astropy/io/fits/hdu/compressed/compressed.py:
```diff
@@ -364,6 +366,28 @@ def __init__(
+                    # use most negative value as default
+                    self._header["BLANK"] = -(1 << (self._bitpix - 1))
+                    warnings.warn(
+                        "Setting default value for missing BLANK keyword in compressed extension",
```
Shouldn't this warning also tell *what* default value was set so it can more easily be fixed upstream by users who also control the fits-writing code ?

neutrinoceros reviewed (commented):
First of, I want to point out this seems like great work to me, though again I'm no authority when it comes to `io.fits`, but your work is greatly appreciated.

Since you added tests (which is great) alongside some binary files, it would be important to also include the code that was used to generate them if possible. See
https://github.com/astropy/astropy/issues/16371 for context.

rlwastro commented:
> First of, I want to point out this seems like great work to me, though again I'm no authority when it comes to `io.fits`, but your work is greatly appreciated.
> 
> Since you added tests (which is great) alongside some binary files, it would be important to also include the code that was used to generate them if possible. See #16371 for context.

@neutrinoceros @pllim 
I can do that, and I agree it seems like a good idea.  But I don't know where to put the script.  I tried putting it in the `data/` directory, but the pre-commit check complains that it needs an `__init__.py` file.  I'm pretty sure we don't want that.  I could put it in `tests/`, but then pytest might try to run it?  I poked around in the astropy directories and did not see any obvious place to use for this sort of thing.

I have the code ready to check in but need some guidance before I can complete the process.

rlwastro commented:
@pllim @neutrinoceros 

Maybe I should try to be clearer about my question: if I have a script `build_my_test_data.py` that creates the files used for testing, should I just stick it in the `tests/` directory?

I do not want it to run as part of testing, but I assume it should be somewhere in the vicinity of the `test` directory since that is what it is for.

dhomeier commented:
Thanks for your work on this! Since saving the code to generate binary data has only been declared a desired policy in #16371, but no decision been made how to put that policy into effect, probably needs further discussion where to put those scripts.
However looking at the test from the [previous attempt to fix this this](https://github.com/astropy/astropy/pull/15756/files#diff-3d706f4a6b686b1787aeb0b2332764235716d20699794cb78f2b0709bd059582) I am wondering if _some_ of the test files could not be created directly in and by the test? As I understand, some of your examples are technically invalid FITS, so probably will not be possible to create (easily) with `io.fits`, but the remainder should.
Still trying to bring this into context with the definition of `BLANK` in the standard; probably will still take someone more familiar with its usage and especially the `ZBLANK` and table columns cases.
Last but not least note that much of this code is scheduled for a significant refactor in #15474, so it might better be hold until that PR is merged.

neutrinoceros commented:
Agreed with @dhomeier that generating data as part of the test seems preferable if it's quick enough.

rlwastro commented:
@dhomeier @neutrinoceros 
The files are generated by running fpack (plus some modifications to the fpack file using astropy.io.fits).  My opinion is that it is much better to rely on a file generated externally for the test than on generating a file on-the-fly.  Changes in the code could result in an error during compression that is compensated by an error in uncompression.  It seems dangerous to assume that cannot happen in this kind of code.

I could modify the tests so that there is only one additional test file.  I would be very reluctant to try to get to zero test files though.

I am very familiar with the FITS compression standard since [I wrote it](https://ui.adsabs.harvard.edu/abs/2012arXiv1201.1336W/abstract)!  So I am confident that my interpretation of the standard and the approach I've taken is correct.  Of course that doesn't mean it is not a good idea to find someone else to review it.

If you think the fix for #15474 will happen soon, I can see the benefit of waiting for this one.  I can't tell from the discussion there whether that major change is really imminent or not.  It is not obvious whether the discussion in that pull request is converging.  It's been 3 months since the request got switched back to a draft.

neutrinoceros commented:
> If you think the fix for https://github.com/astropy/astropy/pull/15474 will happen soon

@astrofrog and @saimn would be in a better position to answer that question

rlwastro commented:
After further thought, I decided that it would be OK to replace the test files with files that are completely generated within the `test_tiles_compression.py` script.  That makes the changes to the test script much longer, but now the additional FITS files are not part of the pull request.

Incidentally, in response to the comment by @dhomeier : none of the compressed files that are tested are actually invalid, as far as I can see.  The FITS spec is reluctant to say _must_ instead of _should_ for any of its requirements, so these all fit in the category of formats that are not the preferred approach but nonetheless are acceptable.  I think that makes them worth testing. It is quite possible that `astropy.io.fits` is more robust to these variations than `cfitsio` and `fpack/funpack` are.

neutrinoceros commented on astropy/io/fits/hdu/compressed/tests/test_tiled_compression.py:
```diff
@@ -65,6 +66,103 @@ def test_zblank_support(canonical_data_base_path, tmp_path):
+            assert hdul[1].header.get("BLANK") == -32768
+
+    # this one generates an expected warning
+    with pytest.warns(AstropyUserWarning) as record:
```
going full `record` style is only needed in cases where more than one warning is expected, but here we should simplify as
```suggestion
    with pytest.warns(
        AstropyUserWarning, 
        match="Setting default value -32768 for missing BLANK keyword in compressed extension",
    ):
```

neutrinoceros commented on astropy/io/fits/hdu/compressed/tests/test_tiled_compression.py:
```diff
@@ -65,6 +66,103 @@ def test_zblank_support(canonical_data_base_path, tmp_path):
+            assert_equal(hdul[1].data, reference)
+            assert hdul[1].header.get("BLANK") == -32768
+
+    unique_warnings = {(wm.category, wm.message.args[0]) for wm in record}
+    expected = {
+        (
+            AstropyUserWarning,
+            "Setting default value -32768 for missing BLANK keyword in compressed extension",
+        ),
+    }
+    assert (
+        unique_warnings == expected
+    ), f"Got some unexpected warnings\n{unique_warnings - expected}"
```
All of this becomes unnecessary with my suggestion above
```suggestion
```

rlwastro replied:
This is a big improvement -- thanks for showing me the right way to do it!

neutrinoceros commented on docs/changes/io.fits/16550.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fix bug for integer compressed images with blanks.
```
```suggestion
Fix a spurious exception when reading integer compressed images with blanks.
```

rlwastro replied:
It is in fact a bug, but I'll accept this anyway!

neutrinoceros replied:
indeed it is, but it'll be marked as such anyway in the final changelog since the file name contains 'bugfix' :)

neutrinoceros commented:
@rlwastro we'll need to rebase this branch as some point to fix docs builds. Are you comfortable with the process ?

rlwastro commented:
> @rlwastro we'll need to rebase this branch as some point to fix docs builds. Are you comfortable with the process ?

@neutrinoceros I think I can probably figure it out.  I struggled last time but pllim helped.

dhomeier reviewed (commented):
Thanks for the clarification and reworking the tests @rlwastro! As mentioned before, this would best get another look from a reviewer more familiar with compressed FITS usage than myself (and @neutrinoceros I suppose). I added some comments after digging a bit deeper into the standard.

> Currently if we read a FITS file with an _uncompressed_ integer image that has a `BLANK` keyword in the header, the image gets converted from an integer datatype to a `float32`, and the blanked pixels are set to `np.nan`. I could argue with that behavior (perhaps a masked integer array would be better than a `float32` array). But that is the way it acts now, and as far as I know people are OK with it.
 
This is not true in all cases; currently uncompressed images that have been encoded as unsigned integers are kept in (the unsigned counterpart of) their integer type even with `BLANK` set, so the blank needs to be applied manually as a mask.
Needless to say, I find this completely confusing, illogical and inconsistent. The standard does not prescribe any particular way how to represent undefined pixels after read-in, and there can be arguments found for both ways, as this is more at the front-end level. But treating it one way (conversion to float and NaN) for plain integers, and the other (preserving int and `BLANK` values) for unsigned-encoded integers just makes no sense to me. All the more since the mask would be fairly easily produced as
`hdu.data == hdu._blank`
in the former case, while the latter, that actually requires it, needs
`hdu.data == hdu._blank + hdu._bzero`

> With my version of the code, a FITS file with a _compressed_ integer image that has a `BLANK` keyword in the header does _not_ get converted to a float. Instead it is kept as an integer array, with the blank values retained as the special integer value.

As you may guess from the example above, I'd actually prefer this behaviour throughout. But as you noted above, conversion to float is the way it acts now (for half the integer types), and changing the default would surely create confusion and in any case constitute a breaking change.
 
> I think for consistency with the uncompressed image behavior, this should instead convert the array to `float32` and replace the blanks with `np.nan`. If you agree, I can withdraw this pull request and submit a new one. But let me know if you think it needs more discussion before doing that.

Probably need to let the `io.fits` maintainers chime in here as well, but at the bugfix level (to get this PR in as soon as possible), I think it should reproduce the current behaviour of uncompressed images, i.e.
1. convert (signed) int to (float) `_dtype_for_bitpix` and set blanks to `np.nan` if BZERO=0, BSCALE=1, [Z]BLANK not None
2. keep pseudo-ints as unsigned `_dtype_for_bitpix` for the respective BZERO values, and leave pixels of BLANK at their values

It may be good to introduce a new argument for consistently making or not making these conversions, and perhaps even add a third option to return the masked integer version, but probably beyond the scope of this PR.
The existing `ignore_blank` would do this at least in one direction, preserving regular integers just like the unsigned ones, so we could consider modifying its defaults and precise meanings.

dhomeier commented:
> Incidentally, in response to the comment by @dhomeier : none of the compressed files that are tested are actually invalid, as far as I can see. The FITS spec is reluctant to say _must_ instead of _should_ for any of its requirements, so these all fit in the category of formats that are not the preferred approach but nonetheless are acceptable. I think that makes them worth testing. It is quite possible that `astropy.io.fits` is more robust to these variations than `cfitsio` and `fpack/funpack` are.

If it can create them, all the better, as it is by default strict on writing – but something that just _should_ not be in there should indeed at most raise a warning. If you feel that something needs to be tested independently of the `io.fits` writer, it is definitely acceptable to add new files in `tests/data`, but I don't see any obvious case here, so the default should be to generate those in the test run (there are a handful of files stored there, but some 100 created by the test suite).

rlwastro commented on astropy/io/fits/hdu/compressed/compressed.py:
```diff
@@ -539,12 +563,12 @@ def _update_header_data(
 
-            if "BLANK" in self._header:
-                blanks = data == np.array(self._header["BLANK"], dtype="int32")
+            if self._blank is not None:
+                blanks = data == np.array(self._blank, dtype="int32")
```
I agree, this is very weird. I haven't changed this part of the code (except to execute it when `BLANK` is set, even without `BSCALE/BZERO`).  I think it would make more sense to do the comparison of data to the blank value before the conversion to float. I don't have any insight into the reason for using `int32` for the datatype.

Maybe this is related to the `uint` thing, which is a can of worms.  I see that `_dtype_for_bitpix()` returns a `uint` type for the special values of `bscale=1` and `bzero=2**bits-1`.  So the type conversion may be changing `data` from `int` to `uint` (rather than to `float`).  The `data` array will stay as type `uint` unless there are blanks, in which case the `np.where(blanks, np.nan, data)` will convert it to a float type.  (I did test this to make sure it works.)

It appears to me this code is simply wrong for `uint` datatypes.  The comparison to the `BLANK` value needs to be done **before the conversion to `uint`**.  If it is done after the conversion to `uint` the results will _not_ be correct.

I have not looked at the `uint` part of the code at all.  I will probably try to fix this since I need a version of this code that works now (for access to Pan-STARRS images, which die with an exception). But if it does not get integrated into a release before #15474 is committed, it will doubtless have to be reworked after that.  I imagine the handling of `BLANK` keywords for unsigned integers in `CompImageHDU` is correct (at least I hope so!) and is different than the code here.

dhomeier replied:
> It appears to me this code is simply wrong for `uint` datatypes. The comparison to the `BLANK` value needs to be done **before the conversion to `uint`**. If it is done after the conversion to `uint` the results will _not_ be correct.

You're right, I did not even notice although I quoted the comparison including `BZERO` elsewhere – but clearly the sensible way to do this would be comparing before conversion. Plain bug to me.

rlwastro commented on astropy/io/fits/hdu/compressed/compressed.py:
```diff
@@ -364,6 +366,28 @@ def __init__(
+        ):
+            # check for column named "ZBLANK"
+            for i in range(1, self._header["TFIELDS"] + 1):
+                if self._header[f"TTYPE{i}"] == "ZBLANK":
```
> Or is the header value here the one that the masked data need to be set to after decompression (which _should_ be `BLANK` according to the standard, but is permitted to be taken from `ZBLANK` instead)? 

That is correct.  `ZBLANK` can be different for each tile, in which case there is no need for a `ZBLANK` value in the header.  But after decompression, if there are blank values in the integer image, there must be a `BLANK` keyword in the uncompressed header that specifies what value should be used for blank pixels.  After uncompression, that blank value has to be the same for the whole image (since there is no memory of which pixels were in different tiles).

I think this an error in the FITS standard.  It should be required to have a `BLANK` value specified for integer images if `ZBLANK` is used.  Otherwise the reader must either invent a value for `BLANK` or must just ignore blanks.  (Getting the FITS standard changed is much harder than changing `astropy` though!)

> In that case this check seems correct, except that the standard does not require `[Z]BLANK` to be present if there actually are no undefined pixel in the image. Maybe having a `ZBLANK` column then is confusing, but it still seems like an allowed usage. But to confirm this, one would need to check if `blank_mask` in `decompress_image_data_section` is empty or not...

In principle this is true.  However, we must be able to generate the header for the uncompressed image without actually reading all the compressed data and scanning it for blanks.  That would be a huge performance hit (e.g., for the `section` functionality, which reads only a part of the compressed data).  So the only knowledge we have is whether `ZBLANK` was specified or not, and the only possible response if we find `ZBLANK` is to include a value for `BLANK`.

dhomeier replied:
Yes, I expected this would not be feasible at this stage, but also thought that if there are no undefined pixels no data would actually be changed anyway. The only catch is that when picking the default blank value here, some values _might_ match it by accident, but since this can be considered a broken, or at least poorly setup file, this should be acceptable. The warning is there to alert the user to the problem.

dhomeier commented:
> I could try to replicate the weird discrepancy between the `uint` and `int` behavior, but that seems like a bad idea because (1) it appears to be wrong and (2) there are big changes coming anyway from changing the `CompImageHDU` to inherit from `ImageHDU` instead of `BinTableHDU`. It does not seem worthwhile to try to do any behavior cleanup of this now.

I tend to agree with (1) now, and quite possibly this combination of `uint` and `BLANK` simply has not been used by anyone since I have seen no questions about the weird behaviour. And the doc clearly states that for `ignore_blank=False` blank values _will_ be converted to NaN, so not doing so in this case can be considered a bug as well.
It seems that #15474 is relegating most or all of this scaling to the functions in [image.py](https://github.com/astropy/astropy/pull/15474/files#diff-1daacadfe9276932278eb86f8b21a7642fb2560cf5408d3e64e2f5ee5dff48c1), this should automatically ensure more consistent behaviour between uncompressed and compressed data. But this will only come in with a major release due to the API changes, so I agree now to bring this bug fix in as soon as possible, and then probably fix the corresponding `image.py` code so everything will still be consistent afterwards.

pllim commented:
Hi. Just catching up on notifications and saw https://github.com/astropy/astropy/pull/16550#issuecomment-2174532227 but not sure if it has been resolved or not.

dhomeier commented:
> Hi. Just catching up on notifications and saw [#16550 (comment)](https://github.com/astropy/astropy/pull/16550#issuecomment-2174532227) but not sure if it has been resolved or not.

There are no new test files added anymore, so no need for external scripts at this point. I want to take a closer look at the current test coverage, but after rebase this looks about ready to go for me.

dhomeier commented on astropy/io/fits/hdu/compressed/tests/test_tiled_compression.py:
```diff
@@ -65,6 +66,95 @@ def test_zblank_support(canonical_data_base_path, tmp_path):
+
+    # create the compressed file
+    cfile1 = tmp_path / "compressed_with_blank.fits"
+    w = np.where(np.isnan(data))
+    blank = -32768
+    data[w] = blank
```
Is not `data` still an integer array at this point? So `np.isnan(data)` should be empty and no `blank`s added to those already set in L90, or am I missing something here?

rlwastro replied:
Absolutely correct, as you mentioned above those lines are completely redundant and don't do anything!  I deleted them in the just-submitted version.

dhomeier reviewed (commented):
Apart from my question on the setup of the data (it seems L95-L98 may be simply redundant), the tests look good to me, thanks for the added coverage!
Of course the uint case could still be tested (or one might even consider to `parametrize` the entire test over all integer dtypes, but that's perhaps overkill).

pllim commented:
Regular CI is all green. But if maintainers also want to run optional CI, please apply `Extra CI` label too. Otherwise, I think this is ready for final review as long as we agree to use the "Squash and Merge" button. Thanks!

rlwastro commented:
> Regular CI is all green. But if maintainers also want to run optional CI, please apply `Extra CI` label too. Otherwise, I think this is ready for final review as long as we agree to use the "Squash and Merge" button. Thanks!

"Squash and Merge" is fine with me.  Thanks!

dhomeier reviewed (approved):
Thanks, everything looking green now, but I'll leave merging to a maintainer.
If you prefer to squash yourself, you can also give `git rebase -i upstream/main` a try; might be reasonable to squash this into 3 or 4 commits.

saimn commented on astropy/io/fits/hdu/compressed/_tiled_compression.py:
```diff
@@ -465,7 +467,11 @@ def decompress_image_data_section(
                     tile_data = tile_data.copy()
-                tile_data[blank_mask] = np.nan
+                if zbitpix > 0:
+                    blank = zblank_header
```
should probably be `= zblank`, which can be `= zblank_header` or `= zblank_column[row_index]` if ZBLANK is specified per tile (see above).

rlwastro replied:
No, it cannot be `zblank_header` because that can be different for every tile.  Once the image has been uncompressed, there must be a single `BLANK` value that is used for the entire image.  There is no longer any memory of the tiles, so using different blank values for different tiles is impossible.

saimn replied:
Right :+1: 
But then we should make sure that the `blank` value here (`zblank_header` for non floating data) is defined. If `ZBLANK`is specified as a column and not in the header, currently you will get None...

rlwastro replied:
There is code in `compressed.py` that adds a `BLANK` keyword to the header (with a warning) for the only case where `BLANK` and `ZBLANK` are missing but `BLANK` is still needed.  (That can occur if there is a `ZBLANK` column in the compressed data table.)  So it is not necessary to test further for it here. If we added redundant code here, I think the test coverage would fail since there would be some code lines that cannot be reached by any tests.

saimn replied:
Sorry I don't see where, can you point me to it ?
And the ZBLANK column case is not tested (which is not surprising, astropy cannot write such a file).

dhomeier replied:
It is this one
https://github.com/astropy/astropy/blob/5c794f66638443263bcc4a04a87b7c438327a8c7/astropy/io/fits/hdu/compressed/compressed.py#L372-L377
and I think the case you mean is `cfile3` in the added test, manually constructed from the columns – one could argue though about testing the case _with_ a (custom) `BLANK` in the header as well.

rlwastro replied:
Thanks @dhomeier -- you are correct.  I included the `cfile3` test to get full coverage of the new code.  I suppose we could do additional tests if you think they would be useful.  They aren't needed for the code coverage though.

If I were to change this, I would probably choose a different value of `BLANK` for the tests that specify it (currently I'm using -32768, which is the same as the default that is supplied by the above code).

dhomeier replied:
> If I were to change this, I would probably choose a different value of `BLANK` for the tests that specify it (currently I'm using -32768, which is the same as the default that is supplied by the above code).

Yes, that would be exactly how I had it in mind, to confirm it is honouring the header value, and not just to keep codecov happy.

dhomeier replied:
Thanks! It had occurred to me that we should also test if the actual `data` for the invalid pixels are set to the `BLANK` value before conversion to float, but the compressed version does not have an `ignore_blank` option, and changing this is beyond the scope of this PR. Should keep that in mind for #15474 though (which may already bring it in automatically from `image.py`).
@pllim does the changelog check mean the changelog file has to be renamed or moved somewhere else, or do we simply ignore or `skip changelog-checks` for backports? I have unchecked the corresponding box in the list, but its meaning is not completely clear to me.

saimn replied:
Oh missed that somehow, sounds good then, thanks!

saimn commented on astropy/io/fits/hdu/compressed/compressed.py:
```diff
@@ -558,7 +582,10 @@ def _scale_data(self, data):
+                # use float32 version of nan to reduce data size for uint conversion
+                # result will still be float64 for larger uint sizes (e.g., uint32)
+                # for float types np.where retains the type of the data array
+                data = np.where(blanks, np.float32(np.nan), data)
```
All this file is being heavily refactored in #15474, and things should be cleaner after since it should inherit more of ImageHDU's behavior... and probably get the float/NaN conversion for free.

rlwastro replied:
I don't doubt that (and certainly hope it is true!)  There might still be some work needed in the #15747  branch to handle unsigned ints consistently.  They are currently not handled correctly in `ImageHDU`.  So chances are you will inherit that incorrect behavior without some additional work on the `ImageHDU` class.

dhomeier replied:
Yes; I'd actually argue for a follow-up PR in this branch to make the corresponding fixes in `image.py`, so 6.1.x will have consistent behaviour of uncompressed files, and #15747 can inherit the fixed code right away.

saimn reviewed (commented):
Thanks for the PR and sorry for the delay (long meeting last week and other things). 
Given the conflicts with #15474 I think we need to wait before merging. Sorry but #15474 is a huge step and will simplify a lot of things including for this PR I think.

astrofrog commented:
I would like to make a suggestion - this is clearly an important bug to fix, but if we wait until after #15474 then any bug fix merged after to the compressed FITS code won't be backportable anyway. Therefore one option would be to change this PR to be against the v6.1.x branch directly, and then once #15474 is merged we can cherry pick the regression test and re-do a new fix for the bug in v7.0.x?

rlwastro commented:
@astrofrog @saimn @dhomeier 

> I would like to make a suggestion - this is clearly an important bug to fix, but if we wait until after #15474 then any bug fix merged after to the compressed FITS code won't be backportable anyway. Therefore one option would be to change this PR to be against the v6.1.x branch directly, and then once #15474 is merged we can cherry pick the regression test and re-do a new fix for the bug in v7.0.x?

I would be willing to give that a try if it seems like an acceptable solution.  (This is all pushing the limits of my comfort with git, but it's good to learn new things.)  I won't try it unless there is agreement that this would be a good approach.

astrofrog commented:
@saimn - what do you think?

astrofrog commented:
An alternative is that we do merge it into main but I would add a revert as my first commit in #15474 and re-implement a fix at the end of the git history in that PR as actually rebasing would be very tricky.

neutrinoceros commented:
FWIW I was about to make the exact same suggestion Tom did in https://github.com/astropy/astropy/pull/16550#issuecomment-2189916395

saimn commented:
Having the bug fixed would be great so yes making the PR against the v6.1.x branch seems a good idea.

neutrinoceros commented:
@rlwastro here are some instructions for this operation:
```
git remote add upstream https://github.com/astropy/astropy.git
git fetch upstream v6.1.x
git checkout rlwastro-compressionFix
git rebase -i v6.1.x
# ... rebasing ...
git push --force
```

Then on this page, click the "Edit" button next to the PR title: you should be able to change the target branch from main to v6.1.x (the dropdown should be next to "open")

rlwastro commented:
I squashed everything into a single commit and changed the fix to be against the `v6.1.x` branch.  Thanks @neutrinoceros for the helpful tips!  I *hope* everything was done correctly. The files changed look correct.  There is a check failure for the change log file.  I'm not sure whether that is something I need to fix or not.

nstarman commented on astropy/io/fits/hdu/compressed/_tiled_compression.py:
```diff
@@ -465,7 +467,11 @@ def decompress_image_data_section(
                 if not tile_data.flags.writeable:
                     tile_data = tile_data.copy()
-                tile_data[blank_mask] = np.nan
+                if zbitpix > 0:
+                    blank = zblank_header
+                else:
+                    blank = np.nan
```
```suggestion
                blank = zblank_header if zbitpix > 0 else np.nan
```

nstarman replied:
Given the shortening, this might also be combined with the following line

rlwastro replied:
OK, made that change too.  I'm not crazy about the 1-line style for assignments but it is fairly readable in this case.

pllim commented:
I haven't been following the convo. The failure isn't from change log, @dhomeier . It is because this PR is opened against v6.1.x and not main. If that is intentional, then close and re-open the PR to re-run the checks.

So, this means that this patch is only going into the next v6.1 series and WILL NOT be in v7.0.0. Is that correct?

dhomeier commented:
Yes, it is intentional – see https://github.com/astropy/astropy/pull/16550#issuecomment-2189916395 for the rationale.
Apparently needed reopening to make the pre-CI checks aware of the changed target, thanks!

saimn reviewed (approved):
Thanks for the fix @rlwastro !

dhomeier commented:
Thank you for your contribution and your patience throughout our review process, @rlwastro! FITS code is something we approach only with great care and caution. ;-)
Thanks also to everyone who helped get this into Astropy 6!
